### ISSUE TITLE
SQLite ResultSet close() blocks Finalizer Queue

### ISSUE THREAD

**Issue by tildedave** [xerial/sqlite-jdbc#349]
Hi, I've noticed an issue with the SQLite ResultSet close() that was added as part of https://github.com/xerial/sqlite-jdbc/pull/215

I'm using xerial-sqlite 3.20.1. 

I have a thread that's running and taking a long time (looks like its not going to return which is its own issue).  However, while it's doing this `CoreResultSet.close()` call finalizer queue is blocked on the `db` object's lock because it's executing the query.  Code here: https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/core/CoreResultSet.java#L130

We have a number of different threads which each own their own SQLite connection.  This synchronized lock causes the whole service to stop finalizing any other objects, which eventually causes the whole service to die.

```
Thread 25408: (state = IN_NATIVE)
 - org.sqlite.core.NativeDB.step(long) @bci=0 (Compiled frame; information may be imprecise)
 - org.sqlite.core.DB.execute(org.sqlite.core.CoreStatement, java.lang.Object[]) @bci=108, line=811 (Compiled frame)
 - org.sqlite.core.CoreStatement.exec() @bci=47, line=75 (Compiled frame)
 - org.sqlite.jdbc3.JDBC3Statement.execute(java.lang.String) @bci=39, line=61 (Compiled frame)
```

```
Thread 8205: (state = BLOCKED)
 - org.sqlite.core.CoreResultSet.close() @bci=49, line=122 (Compiled frame)
 - org.sqlite.jdbc4.JDBC4ResultSet.close() @bci=6, line=34 (Compiled frame)
 - org.sqlite.jdbc3.JDBC3ResultSet.finalize() @bci=1, line=163 (Compiled frame)
 - java.lang.System$2.invokeFinalize(java.lang.Object) @bci=1, line=1270 (Compiled frame)
 - java.lang.ref.Finalizer.runFinalizer(sun.misc.JavaLangAccess) @bci=46, line=98 (Compiled frame)
 - java.lang.ref.Finalizer.access$100(java.lang.ref.Finalizer, sun.misc.JavaLangAccess) @bci=2, line=34 (Compiled frame)
 - java.lang.ref.Finalizer$FinalizerThread.run() @bci=45, line=213 (Compiled frame)
```

Is there a recommended way to fix this?  One thought is that I can try to adjust my code to avoid this issue by keeping all the ResultSet objects in scope until the end of the thread's life, so `finalize` will never be run in this scenario.

Thanks much for your thoughts on this!

**Comment by tildedave** [comment]
Okay, so I moved all my ResultSet objects to be retained by the class that holds the SQLiteConnection object so I wouldn't run into this again, seems to be working.  (No service deaths since deployed, at least.)

One other thing I noticed when debugging this is that `close()` will still try to clean things up if you call it multiple times.  Our ResultSet objects are always used in a try-with-resources block which means that the bug I ran into was with objects that were already closed.

So, question: Could `CoreResultSet` bail out in a matter similar to how CoreStatement works? https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/core/CoreStatement.java#L118  Seems like this avoids synchronizing in the double-close scenario.

**Comment by b3nn0** [comment]
Running into the same issue from time to time. Sadly, I can't easily move the ResultSet destruction to other places. Also, I'm not sure what triggers this, as I wasn't able to reproduce it myself.

About bailing out of the close: I'm all for it. But I'm not sure if it will really fix the issue.

**Comment by tildedave** [comment]
@b3nn0 Cool, I eventually circled back here and submitted a patch.
